resolver.go

```diff
@@ -314,3 +327,30 @@ func (r *Resolver) DialContext(ctx context.Context, network, addr string) (net.C
 	}
 	return nil, lastErr
 }
+
+func (r *Resolver) AddZone(name string, zone Zone) error {
+	r.zonesMutex.Lock()
+	defer r.zonesMutex.Unlock()
+
```

Comment:
Would be nice if this canonicalized the name, since that's what the server tries to match during lookups.
```suggestion
	name = dns.CanonicalName(name)
```